tests/test_package_swap.py

```diff
@@ -0,0 +1,262 @@
+
+
+class TestPackageSwap(common.BaseUDES):
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestPackageSwap, cls).setUpClass()
+        User = cls.env['res.users']
+
+        user_warehouse = User.get_user_warehouse()
+        cls.picking_type_pick = user_warehouse.pick_type_id
+        cls.pack_4apples_info = [{'product': cls.apple,
+                                  'qty': 4}]
+
+    def setUp(self):
+        super(TestPackageSwap, self).setUp()
+        Package = self.env['stock.quant.package']
+
+        self.picking_type_pick.u_reserve_as_packages = True
+        self.picking_type_pick.u_allow_swapping_packages = True
+        self.expected_package = Package.get_package(EXPECTED_PACKAGE_NAME,
+                                                    create=True)
+        self.scanned_package = Package.get_package(SCANNED_PACKAGE_NAME,
+                                                   create=True)
+
+    def test01_error_if_not_allowed_to_swap(self):
+        """ Should error if the 'allow swap' attr is unflagged """
+        self.picking_type_pick.u_allow_swapping_packages = False
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.expected_package.id)
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.scanned_package.id)
+        picking = self.create_picking(self.picking_type_pick,
+                                      products_info=self.pack_4apples_info,
+                                      confirm=True,
+                                      assign=True)
+
+        with self.assertRaises(ValidationError):
+            picking.update_picking(package_name=SCANNED_PACKAGE_NAME,
+                                   expected_package_name=EXPECTED_PACKAGE_NAME)
+
+    def test02_reserved_and_unreserved(self):
+        """
+        Should error when swapping two packages:
+         - with same contents;
+         - in the same location;
+         - the scanned package is reserved, the expected one not.
+
+        """
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.scanned_package.id)
+        picking = self.create_picking(self.picking_type_pick,
+                                      products_info=self.pack_4apples_info,
+                                      confirm=True,
+                                      assign=True)
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.expected_package.id)
+
+        # checking pre-conditions
+        self.assertTrue(self.scanned_package.is_reserved())
+        self.assertFalse(self.expected_package.is_reserved())
+
+        with self.assertRaises(ValidationError):
+            picking.update_picking(package_name=SCANNED_PACKAGE_NAME,
+                                   expected_package_name=EXPECTED_PACKAGE_NAME)
```

Comment:
are we testing that the expected package is not found in the picking?
Please add the check of the message, this will also help to check what we are testing.

Reply:
Hmmm, not sure if I agree here.
The cause of the error is what your mentioning, but what we're checking is the actual error. We shouldn't be interested in the cause. The value of this test is to prevent future behaviour changes with respect of this assumption (if the expected package is not reserved, `updated_picking` fails). If you start making assumptions on the whole stack, this test becomes pointless as it would be equivalent of testing udes_core.

Reply:
We are testing that the expected package is not found in the picking, because if the expected package is reserved for a different picking update_picking will also raise the same error.

Reply:
Yep, you're right, the message doesn't come from the parent class.